src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserRepository.cs

```diff
@@ -24,4 +52,180 @@ public async Task<User> FindDeletedUserPlaceholder(CancellationToken cancellatio
             cancellationToken
         );
     }
+
+    public async Task<User> CreateOrUpdate(
+        string email,
+        GlobalRoles.Role role,
+        Guid createdById,
+        DateTimeOffset? createdDate = null,
+        CancellationToken cancellationToken = default
+    ) =>
+        await CreateOrUpdate(
+            email: email,
+            roleId: role.GetEnumValue(),
+            createdById: createdById,
+            createdDate: createdDate,
+            cancellationToken: cancellationToken
+        );
+
+    public async Task<User> CreateOrUpdate(
+        string email,
+        string roleId,
+        Guid createdById,
+        DateTimeOffset? createdDate = null,
+        CancellationToken cancellationToken = default
+    )
+    {
+        if (createdDate > DateTimeOffset.UtcNow)
+        {
+            throw new ArgumentException($"{nameof(User)} created date cannot be a future date.");
+        }
+
+        var existingUser = await contentDbContext.Users.SingleOrDefaultAsync(i =>
+            i.Email.ToLower().Equals(email.ToLower())
+        );
+
+        return existingUser is null
+            ? await CreateNewUser(
+                email: email,
+                roleId: roleId,
+                createdById: createdById,
+                createdDate: createdDate,
+                cancellationToken: cancellationToken
+            )
+            : await UpdateExistingUser(
+                existingUser: existingUser,
+                roleId: roleId,
+                createdById: createdById,
+                createdDate: createdDate,
+                cancellationToken: cancellationToken
+            );
+    }
+
+    public async Task SoftDeleteUser(User activeUser, Guid deletedById, CancellationToken cancellationToken = default)
+    {
+        contentDbContext.Attach(activeUser);
+
+        activeUser.Active = false;
+        activeUser.SoftDeleted = DateTime.UtcNow;
+        activeUser.DeletedById = deletedById;
+
+        await contentDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private async Task<User> CreateNewUser(
+        string email,
+        string roleId,
+        Guid createdById,
+        DateTimeOffset? createdDate,
+        CancellationToken cancellationToken
+    )
+    {
+        var newUser = new User
+        {
+            Email = email.ToLower(),
+            RoleId = roleId,
+            Active = false,
+            CreatedById = createdById,
+            Created = createdDate ?? DateTimeOffset.UtcNow,
```

Comment:
Just thinking, `createdDate` could theoretically be in any offset here, so should we do something like

```csharp
Created = createdDate?.ToUniversalTime() ?? DateTimeOffset.UtcNow,
```

here , so that we save the value as UTC?

This might be one for discussion to make sure that we save values consistently using UTC everywhere that we create/update data, if we are bothered about `DateTimeOffset` values having a consistent offset.

It could also be one that we need to wait to find our way with as we get more experience of using `DateTimeOffset`.